pkg/syncer/namespace/namespace_downstream_controller.go

```diff
@@ -84,24 +97,46 @@ func NewDownstreamController(
 	kubeClient := kubernetes.NewForConfigOrDie(downstreamConfig)
 
 	c := DownstreamController{
-		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), downstreamControllerName),
-
+		queue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), downstreamControllerName),
+		delayedQueue: workqueue.NewNamedDelayingQueue(downstreamControllerName),
+		toDeleteMap:  make(map[string]time.Time),
 		deleteDownstreamNamespace: func(ctx context.Context, namespace string) error {
 			return downstreamClient.Resource(namespaceGVR).Delete(ctx, namespace, metav1.DeleteOptions{})
 		},
 		upstreamNamespaceExists: func(clusterName logicalcluster.Name, upstreamNamespaceName string) (bool, error) {
 			_, err := upstreamInformers.ForResource(namespaceGVR).Lister().ByCluster(clusterName).Get(upstreamNamespaceName)
-			if errors.IsNotFound(err) {
+			if apierrors.IsNotFound(err) {
 				return false, nil
 			}
-			return !errors.IsNotFound(err), err
+			return !apierrors.IsNotFound(err), err
 		},
 		getDownstreamNamespace: func(downstreamNamespaceName string) (runtime.Object, error) {
 			return downstreamInformers.ForResource(namespaceGVR).Lister().Get(downstreamNamespaceName)
 		},
 		listDownstreamNamespaces: func() ([]runtime.Object, error) {
 			return downstreamInformers.ForResource(namespaceGVR).Lister().List(labels.Everything())
 		},
+		isDowntreamNamespaceEmpty: func(ctx context.Context, namespace string) (bool, error) {
```

Comment:
This seems slightly backwards, since the namespace lifecycle controller in k8s will remove these things during namespace deletion finalizer. Why do we need to invert the flow here?

Reply:
Dowstream namespaces are created at will by the Syncer to recieve resources synced from upstream. But *real syncing happens at the individual resource level*. What happens at the namespace level is *scheduling*, but that's a kcp-only thing.
So the signal to delete a downstream namespace is that no more resources synced from KCP exists for this downstream namespace.
Having the deletion of a downstream namespace be triggered only by the fact that the corresponding upstream namespace does not exist (or more precisely is not visible by the syncer) was precisely the cause of the bug this PR is trying to fix.
The "removal-from-a-synctarget" process is handled at the KCP level, by maintaining consistency between namespace scheduling and resource scheduling. So a downstream namespace should never end up in `Terminating` state because having been deleted though still having synced resources inside it.